File: mocker-back/src/main/java/com/mocker/service/SchemaService.java
package com.mocker.service;

import com.mocker.domain.model.entity.Schema;
import com.mocker.domain.model.entity.Table;
import com.mocker.domain.model.entity.TableRelation;
import com.mocker.domain.model.entity.enumeration.Role;

import java.util.List;
import java.util.UUID;

/**
 * @author Luong Quoc Trung
 * @author Do Quoc Viet
 */

public interface SchemaService {
    List<Schema> getSchemasByProjectId(UUID projectId, List<Role> roles);

    Schema upsert(Schema schema);

    Schema delete(UUID id);

    Schema getSchema(UUID id);

    List<Schema> getSchemas(List<Role> roles);

    List<Table> getTablesBySchema(UUID schemaId);

    List<TableRelation> getTableRelationsBySchema(UUID schemaId);
}

File: mocker-front/src/app/core/service/schema.service.ts
import { Injectable } from '@angular/core';
import { SchemaModel } from '@core/domain/model/entity/schema.model';
import { AbstractService } from '@core/service/abstract.service';
import { Observable } from 'rxjs';
import { TableModel } from '@core/domain/model/entity/table.model';
import { HttpMethodEnum } from '@core/domain/enum/http-method.enum';
import { TableRelationModel } from '@core/domain/model/entity/table-relation.model';
import { RoleEnum } from '@core/domain/enum/role.enum';
import { StringUtil } from '@core/util/string.util';

/**
 * @author Do Quoc Viet
 */

@Injectable({
  providedIn: 'root'
})
export class SchemaService extends AbstractService<SchemaModel> {
  override ENTITY_URL: string = 'schema';

  getSchemasByProjectId(projectId: string): Observable<SchemaModel[]> {
    return this.request<SchemaModel[]>(HttpMethodEnum.GET, `${this.API_URL}/${this.ENTITY_URL}?projectId=${projectId}`);
  }

  getSchemas(roles: RoleEnum[]): Observable<SchemaModel[]> {
    return this.request<SchemaModel[]>(HttpMethodEnum.GET, `${this.API_URL}/${this.ENTITY_URL}?${roles ? `roles=${roles.join(',')}` : StringUtil.EMPTY}`);
  }

  getTablesBySchema(schemaId: string): Observable<TableModel[]> {
    return this.request<TableModel[]>(HttpMethodEnum.GET, `${this.API_URL}/${this.ENTITY_URL}/${schemaId}/tables`);
  }

  getTableRelationsBySchema(schemaId: string): Observable<TableRelationModel[]> {
    return this.request<TableRelationModel[]>(HttpMethodEnum.GET, `${this.API_URL}/${this.ENTITY_URL}/${schemaId}/tables/table-relations`);
  }

}
